test/unit/codegen/codegen_utils.cpp

```diff
@@ -0,0 +1,218 @@
+                float_constant);
+            REQUIRE(nmodl_constant_result == codegen_output);
+        }
+    }
+
+    GIVEN("Double constants in scientific notation as strings") {
+        std::string double_constant_with_e = "1e+18";
+        std::string double_constant_with_e_minus = "1e-18";
+        std::string double_constant_with_decimal_point_e = ".123e18";
+        std::string double_constant_with_E = "1E18";
+
+
+        std::string result_double_constant_with_e = "1d+18";
+        std::string result_double_constant_with_e_minus = "1d-18";
+        std::string result_double_constant_with_decimal_point_e = ".123d18";
+        std::string result_double_constant_with_E = "1d18";
+
+        THEN("Codegen ISPC Visitor prints doubles with scientific notation") {
+            auto nmodl_constant_result = codegen::utils::double_to_string<CodegenIspcVisitor>(
+                double_constant_with_e);
+            REQUIRE(nmodl_constant_result == result_double_constant_with_e);
+            nmodl_constant_result = codegen::utils::double_to_string<CodegenIspcVisitor>(
+                double_constant_with_e_minus);
+            REQUIRE(nmodl_constant_result == result_double_constant_with_e_minus);
+            nmodl_constant_result = codegen::utils::double_to_string<CodegenIspcVisitor>(
+                double_constant_with_decimal_point_e);
+            REQUIRE(nmodl_constant_result == result_double_constant_with_decimal_point_e);
+            nmodl_constant_result = codegen::utils::double_to_string<CodegenIspcVisitor>(
+                double_constant_with_E);
+            REQUIRE(nmodl_constant_result == result_double_constant_with_E);
+        }
+    }
+
+    GIVEN("Float constants in scientific notation as strings") {
+        std::string float_constant_with_e = "1e+18";
+        std::string float_constant_with_e_minus = "1e-18";
+        std::string float_constant_with_decimal_point_e = ".123e18";
+        std::string float_constant_with_E = "1E18";
+
+        THEN("Codegen ISPC Visitor prints doubles with scientific notation") {
+            auto nmodl_constant_result = codegen::utils::float_to_string<CodegenIspcVisitor>(
+                float_constant_with_e);
+            REQUIRE(nmodl_constant_result == float_constant_with_e);
+            nmodl_constant_result = codegen::utils::float_to_string<CodegenIspcVisitor>(
+                float_constant_with_e_minus);
+            REQUIRE(nmodl_constant_result == float_constant_with_e_minus);
+            nmodl_constant_result = codegen::utils::float_to_string<CodegenIspcVisitor>(
+                float_constant_with_decimal_point_e);
+            REQUIRE(nmodl_constant_result == float_constant_with_decimal_point_e);
+            nmodl_constant_result = codegen::utils::float_to_string<CodegenIspcVisitor>(
+                float_constant_with_E);
+            REQUIRE(nmodl_constant_result == float_constant_with_E);
+        }
+    }
+}
```

Comment:
can we test also -0 and +0? Not sure it is so important. Otherwise, very good!

Reply:
`-` and `+` are parsed by the lexer/parser and they are not passed to these functions, so `0` will be handled like `1`